### ISSUE TITLE
Consul : Consul Leader issue

### ISSUE THREAD

**Issue by naveensrinivasan** [helm/charts#1289]
**Is this a request for help?**:
Yes 
---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
Bug Report

**Version of Helm and Kubernetes**:

**Which chart**:
consul

**What happened**:
Consul fails with 
```
2017/06/13 15:37:48 [ERR] agent: failed to sync remote state: No cluster leader
    2017/06/13 15:38:02 [ERR] agent: coordinate update error: No cluster leader
    2017/06/13 15:38:24 [ERR] agent: failed to sync remote state: No cluster leader
    2017/06/13 15:38:31 [ERR] agent: coordinate update error: No cluster leader
    2017/06/13 15:38:49 [ERR] agent: failed to sync remote state: No cluster leader
    2017/06/13 15:39:02 [ERR] agent: coordinate update error: No cluster leader
    2017/06/13 15:39:16 [ERR] agent: failed to sync remote state: No cluster leader
    2017/06/13 15:39:24 [ERR] agent: coordinate update error: No cluster leader
    2017/06/13 15:39:45 [ERR] agent: failed to sync remote state: No cluster leader
```

**What you expected to happen**:
To automatically choose the leader

**How to reproduce it** (as minimally and precisely as possible):
helm install stable/consul  --name=consul

**Anything else we need to know**:
Nothing else.

**Comment by moliver-resgen** [comment]
I am also having this issue. Kubernetes 1.6.4.  Tests fail.

**Comment by nrvale0** [comment]
I'm having this issue and a ```k8s log <name>-0``` is showing that the in a 3 node SS node 0 and node 2 have conflicting node IDs:

```
helm install --name somecustomer --set imagetag=0.8.3,ui.enabled=true,uiService.enabled=true,ConsulDnsPort=53 consul
```

```
    2017/06/16 18:12:35 [INFO] raft: Initial configuration (index=0): []
    2017/06/16 18:12:35 [INFO] serf: EventMemberJoin: somecluster-consul-0 10.44.2.26
    2017/06/16 18:12:35 [INFO] serf: EventMemberJoin: somecluster-consul-0.dc1 10.44.2.26
    2017/06/16 18:12:35 [INFO] raft: Node at 10.44.2.26:8300 [Follower] entering Follower state (Leader: "")
    2017/06/16 18:12:35 [INFO] consul: Adding LAN server somecluster-consul-0 (Addr: tcp/10.44.2.26:8300) (DC: dc1)
    2017/06/16 18:12:35 [INFO] consul: Handled member-join event for server "somecluster-consul-0.dc1" in area "wan"
    2017/06/16 18:12:35 [INFO] agent: Joining cluster...
    2017/06/16 18:12:35 [INFO] agent: (LAN) joining: [10.44.2.26 10.44.1.23 10.44.2.27]
    2017/06/16 18:12:35 [INFO] agent: (LAN) joined: 1 Err: <nil>
    2017/06/16 18:12:35 [INFO] agent: Join completed. Synced with 1 initial agents
    2017/06/16 18:12:35 [INFO] serf: EventMemberJoin: somecluster-consul-1 10.44.1.23
    2017/06/16 18:12:35 [INFO] consul: Adding LAN server somecluster-consul-1 (Addr: tcp/10.44.1.23:8300) (DC: dc1)
    2017/06/16 18:12:35 [INFO] serf: EventMemberJoin: somecluster-consul-1.dc1 10.44.1.23
    2017/06/16 18:12:35 [INFO] consul: Handled member-join event for server "somecluster-consul-1.dc1" in area "wan"
    2017/06/16 18:12:35 [ERR] memberlist: Failed push/pull merge: Member 'somecluster-consul-2' has conflicting node ID 'b59c01d9-6ac2-6bee-7e1e-a821ada3b892' with this agent's ID from=10.44.2.27:35328
    2017/06/16 18:12:35 [WARN] memberlist: ignoring alive message for 'somecluster-consul-2': Member 'somecluster-consul-2' has conflicting node ID 'b59c01d9-6ac2-6bee-7e1e-a821ada3b892' with this agent's ID
```

**Comment by nrvale0** [comment]
It looks to me like this happens because the gopsutil library used to generate the Consul Node ID is based on the underlying node/minions's host ID. If the k8s schedules more than one Consul Pod on the same node/minion then they will have identical Consul Node IDs.

**Comment by nrvale0** [comment]
related : https://github.com/hashicorp/consul/issues/3070

**Comment by nrvale0** [comment]
A temporary work-around might be to ensure the Consul Pods are scheduled with node or pod anti-affinity. I think not all versions of k8s in which folks might try to install this Chart support anti-affinity scheduling though.

**Comment by nrvale0** [comment]
This is not quite working but here is an attempt to work around this problem via scheduler pod anti-affinity:

https://github.com/nrvale0/charts/commits/consul_add_pod_antiaffinity

I suspect it is not working because the docs for pod anti-affinity are ... less than clear.

**Comment by naveensrinivasan** [comment]
@nrvale0 👍  Thanks

**Comment by mitom** [comment]
Consul by default generates the pod id based off of attributes of the host, this results on identical UUIDs if you run multiple consul instances on the same host. There is an option to turn this off and instead generate a random UUID, which you can [set via the chart as well](https://github.com/kubernetes/charts/blob/master/stable/consul/values.yaml#L26-L29).

**Comment by nrvale0** [comment]
@mitom I just cordoned off a bunch of nodes such that it guaranteed that Consul Pods would end up being scheduled on the same node and then deployed with DisableHostNodeID=true and indeed that does resolve the conflicts. (Anti-affinity disabled for this run.)

Does it not make sense that -disable-host-node-id should always be set? I can't think of a reason why not.

**Comment by mitom** [comment]
I can only think of 1 scenario, but I am unsure whether it's justified or not. If a pod were to be removed from a k8s node for whatever reason (like scaling down) and everything cleaned up afterwards (PVCs), when the cluster would scale out again, it could come back with the same node id. I don't know if that's desired or the opposite though.

I am more in favour of setting affinity anyway as it doesn't make much sense to have 2 consul containers on the same k8s node, at that point might as well scale down consul.

Overall I'd say it's easy to set the value, or unset it. Either way, `false` is the consul default for it so I would assume that's the reason why it's set that way.

**Comment by edbergavera** [comment]
Hi @naveensrinivasan @nrvale0 @mitom 

We have been experiencing (at least 3 times already) this same exact issue where consul fails to elect a leader. Unfortunately, we cannot use the info on https://www.consul.io/docs/guides/outage.html as we can see it's not applicable in this situation. What we currently do is to export KVs, purge consul chart, install chart and import back KVs. This doesn't sound a better approach though but that is something that work in our situation.

**Comment by edbergavera** [comment]
@naveensrinivasan 

After my last post above, we encountered again this issue. Any idea about a workaround on this? I can't seem to figure out what causes for not being able to elect a cluster leader.

**Comment by rodrigovidal** [comment]
Same problem here.

**Comment by rajiteh** [comment]
Setting the value `DisableHostNodeId` to  `true` fixes this for me. I was able to run 3 consul Pods in the same Node.

https://github.com/kubernetes/charts/commit/6082ac5fcdd0d2bdf509c890a8265bf398f3a038

**Comment by edbergavera** [comment]
@rajiteh It seemed to work if Consul pods are on the same Node. However, I haven't tried experimenting killing one of the Consul pods to see if still in snyc. Would you be able to verify this on your end also. Let us know :)

**Comment by rajiteh** [comment]
@edbergavera there's a different issue with rescheduled members not being able to join the cluster. I believe your issue is similar to https://github.com/kubernetes/charts/issues/1143 

I have proposed a fix for it here: https://github.com/kubernetes/charts/pull/1556

**Comment by jonstacks** [comment]
@naveensrinivasan, I'm thinking this issue can be closed now since the consul version is now at 1.0.0. Would you mind testing and let me know if you are still having issues?

**Comment by naveensrinivasan** [comment]
Things are working. Thanks.
